// @azure/msal-browser/dist/internals.mjs
/*! @azure/msal-browser v3.0.2 2023-08-18 */
'use strict';
export { BrowserCacheManager } from './cache/BrowserCacheManager.mjs';
export { CacheRecord } from '@azure/msal-common';
export { StandardInteractionClient } from './interaction_client/StandardInteractionClient.mjs';
export { RedirectClient } from './interaction_client/RedirectClient.mjs';
export { PopupClient } from './interaction_client/PopupClient.mjs';
export { SilentIframeClient } from './interaction_client/SilentIframeClient.mjs';
export { SilentCacheClient } from './interaction_client/SilentCacheClient.mjs';
export { SilentRefreshClient } from './interaction_client/SilentRefreshClient.mjs';
export { NativeInteractionClient } from './interaction_client/NativeInteractionClient.mjs';
export { RedirectHandler } from './interaction_handler/RedirectHandler.mjs';
export { EventHandler } from './event/EventHandler.mjs';
export { NativeMessageHandler } from './broker/nativeBroker/NativeMessageHandler.mjs';
export { BrowserConstants, TemporaryCacheKeys } from './utils/BrowserConstants.mjs';
export { CryptoOps } from './crypto/CryptoOps.mjs';
export { NativeAuthError } from './error/NativeAuthError.mjs';

/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
/**
 * Warning: This set of exports is purely intended to be used by other MSAL libraries, and should be considered potentially unstable. We strongly discourage using them directly, you do so at your own risk.
 * Breaking changes to these APIs will be shipped under a minor version, instead of a major version.
 */
// Cache Manager
